python/pyspark/sql/dataframe.py

```diff
@@ -2188,6 +2188,142 @@ def cube(self, *cols: "ColumnOrName") -> "GroupedData":  # type: ignore[misc]
+        """
+        Unpivot a DataFrame from wide format to long format, optionally leaving
+        identifier columns set. This is the reverse to `groupBy(...).pivot(...).agg(...)`,
+        except for the aggregation, which cannot be reversed.
+
+        This function is useful to massage a DataFrame into a format where some
+        columns are identifier columns ("ids"), while all other columns ("values")
+        are "unpivoted" to the rows, leaving just two non-id columns, named as given
+        by `variableColumnName` and `valueColumnName`.
+
+        When no "id" columns are given, the unpivoted DataFrame consists of only the
+        "variable" and "value" columns.
+
+        All "value" columns must share a least common data type. Unless they are the same data type,
+        all "value" columns are cast to the nearest common data type. For instance, types
+        `IntegerType` and `LongType` are cast to `LongType`, while `IntegerType` and `StringType`
+        do not have a common data type and `unpivot` fails.
+
+        :func:`groupby` is an alias for :func:`groupBy`.
+
+        .. versionadded:: 3.4.0
+
+        Parameters
+        ----------
+        ids : str, Column, tuple, list, optional
+            Column(s) to use as identifiers. Can be a single column or column name,
+            or a list for multiple columns.
+        values : str, Column, tuple, list, optional
+            Column(s) to unpivot. Can be a single column or column name, or a list
+            for multiple columns. If not specified or empty, uses all columns that
+            are not set as `ids`.
+        variableColumnName : scalar, default 'variable'
+            Name of the variable column. If None it uses 'variable'.
+        valueColumnName : scalar, default 'value'
+            Name of the value column. If None it uses 'value'.
+
+        Returns
+        -------
+        DataFrame
+            Unpivoted DataFrame.
+
+        Examples
+        --------
+        >>> df = spark.createDataFrame(
+        ...     [(1, 11, 1.1), (2, 12, 1.2)],
+        ...     ["id", "int", "double"],
+        ... )
+        >>> df.show()
+        +---+---+------+
+        | id|int|double|
+        +---+---+------+
+        |  1| 11|   1.1|
+        |  2| 12|   1.2|
+        +---+---+------+
+
+        >>> df.unpivot("id").show()
+        +---+--------+-----+
+        | id|variable|value|
+        +---+--------+-----+
+        |  1|     int| 11.0|
+        |  1|  double|  1.1|
+        |  2|     int| 12.0|
+        |  2|  double|  1.2|
+        +---+--------+-----+
+        """
+        def to_jcols(cols) -> JavaObject:
+            l = cols
+            if cols is None:
+                l = []
+            elif isinstance(cols, tuple):
+                l = list(cols)
+            elif not isinstance(cols, list):
+                l = [cols]
+            return self._jcols(*l)
+
+        if variableColumnName is None:
+            variableColumnName = "variable"
```

Comment:
default values `variableColumnName = "variable"` and `valueColumnName = "value"` are only used in pyspark?

Scala side doesn't have similar logic?

Reply:
> Scala side doesn't have similar logic?

Seems like Scala side doesn't have similar logic according to the [implementation](https://github.com/apache/spark/pull/36150/files) ?

Reply:
The rational in Scala was that other SQL systems require variable and value column names (there is no default), so we decided Scala API does not have default values either. Given that PySpark Pandas API has default values I thought PySpark Dataframe API could have one too.

If you prefer PySpark Dataframe API to match Scala Dataset API rather than PySpark Pandas API, I am fine with that.

Reply:
i think we'd better keep the behaviors the same.

Reply:
removed